### PR TITLE
ledger::VisitData is now defined as mojom

### PR DESCRIPTION
**Pull request by jasonrsadler**
Fixes brave/brave-browser#4693

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

1. Start Brave and enable Rewards
2. Ensure all panel functionality is retained.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by jasonrsadler** [comment]
There are places where VisitData will be passed copy-wise through `std::bind` cc @bridiver

**Comment by jasonrsadler** [comment]
@kylehickinson for iOS

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 1]
```diff
@@ -376,7 +376,7 @@ - (void)publisherActivityFromURL:(NSURL *)URL
 {
   auto visitData = [self visitDataForURL:URL tabId:0];
   visitData.favicon_url = std::string(faviconURL.absoluteString.UTF8String);
-  ledger->GetPublisherActivityFromUrl(1, visitData, std::string(publisherBlob.UTF8String));
+  ledger->GetPublisherActivityFromUrl(1, visitData.Clone(), std::string(publisherBlob.UTF8String));
```
Comment (bridiver) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm]: do we need to clone here? Can't we just std::move?

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 2]
```diff
@@ -714,7 +714,7 @@ - (void)applicationDidBackground
 - (void)reportLoadedPageWithURL:(NSURL *)url tabId:(UInt32)tabId
 {
   const auto visit = [self visitDataForURL:url tabId:tabId];
-  ledger->OnLoad(visit, [[NSDate date] timeIntervalSince1970]);
+  ledger->OnLoad(visit.Clone(), [[NSDate date] timeIntervalSince1970]);
```
Comment (bridiver) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm]: same as above, I think all these can be `std::move`

File: vendor/brave-ios/Ledger/BATBraveLedger.mm (hunk 1 above)
Comment (kylehickinson) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: The returned value is a not a Ptr variant so nothing to move. I have to go back through the code and replace all the base type and Clone() combinations with Ptr and move's instead

Comment (bridiver) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: I see, this is the raw struct and you're converting it back to Ptr

File: vendor/bat-native-ledger/src/bat/ledger/internal/media/twitter.cc [hunk 3]
```diff
@@ -281,14 +281,14 @@ void MediaTwitter::SavePublisherInfo(
     return;
   }
 
-  ledger::VisitData visit_data;
-  visit_data.provider = TWITTER_MEDIA_TYPE;
-  visit_data.url = url;
-  visit_data.favicon_url = favicon_url;
-  visit_data.name = publisher_name;
+  ledger::VisitDataPtr visit_data = ledger::VisitData::New();
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/media/twitter.cc]: Why do we need to do a pointer here? Why not same as we do it here https://github.com/brave/brave-core/pull/2581/files#diff-6904081a34e36c0d05014066ffff4775R324?

File: vendor/bat-native-ledger/src/bat/ledger/internal/bat_get_media.cc [hunk 4]
```diff
@@ -41,44 +42,44 @@ std::string BatGetMedia::GetLinkType(const std::string& url,
...
-                               const ledger::VisitData& visit_data) {
-  if (parts.size() == 0 || !ledger_->GetRewardsMainEnabled()) {
+                               ledger::VisitDataPtr visit_data) {
+  if (parts.size() == 0 || !ledger_->GetRewardsMainEnabled() || !visit_data) {
     return;
   }
 
   if (type == YOUTUBE_MEDIA_TYPE) {
-    media_youtube_->ProcessMedia(parts, visit_data);
+    media_youtube_->ProcessMedia(parts, *visit_data);
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/bat_get_media.cc]: Why wouldn't we use pointer at this point as well? Why conversion?

File: vendor/bat-native-ledger/src/bat/ledger/internal/media/twitch.h [hunk 5]
```diff
@@ -48,12 +48,12 @@ class MediaTwitch : public ledger::LedgerCallbackHandler {
   static std::string GetMediaURL(const std::string& mediaId);
 
   static std::string GetTwitchStatus(
-    const ledger::TwitchEventInfo& old_event,
-    const ledger::TwitchEventInfo& new_event);
+      const ledger::TwitchEventInfo& old_event,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/media/twitch.h]: we should avoid doing this. If you want to do formatting do it in a separate PR as this introduce changes that are not related to this PR

File: vendor/bat-native-ledger/src/bat/ledger/internal/media/twitter.cc (hunk 3 above)
Comment (bridiver) [reply vendor/bat-native-ledger/src/bat/ledger/internal/media/twitter.cc]: it eventually needs to be converted to VisitDataPtr so why wouldn't we just create it that way?

File: vendor/bat-native-ledger/src/bat/ledger/internal/bat_get_media.cc (hunk 4 above)
Comment (bridiver) [reply vendor/bat-native-ledger/src/bat/ledger/internal/bat_get_media.cc]: the way these to `if` statements are structured with return statements is very strange. If we had an if/else it would be much safer to std::move the visit_data without worrying about someone changing it in a way that attempted to access it after move
